core/src/bls_sigverify/bls_sigverifier.rs

```diff
@@ -0,0 +1,1675 @@
+                })
+        })
+        .try_reduce_with(|mut a, b| {
+            a.aggregate_with(std::iter::once(&b))
+                .map_err(|_| CertVerifyError::KeyAggregationFailed)?;
+            Ok(a)
+        })
+        .ok_or(CertVerifyError::KeyAggregationFailed)?
+}
+
+#[derive(Debug, Error, PartialEq)]
+enum CertVerifyError {
+    #[error("Failed to find key to rank map for slot {0}")]
+    KeyToRankMapNotFound(Slot),
+
+    #[error("Failed to decode bitmap {0:?}")]
+    BitmapDecodingFailed(DecodeError),
+
+    #[error("Failed to aggregate public keys")]
+    KeyAggregationFailed,
+
+    #[error("Failed to serialize original vote")]
+    SerializationFailed,
+
+    #[error("The signature doesn't match")]
+    SignatureVerificationFailed,
+
+    #[error("Pubkey not found for rank {0}")]
+    PubkeyNotFoundForRank(usize),
+
+    #[error("Invalid Pubkey for rank {0}")]
+    InvalidPubkeyForRank(usize),
+
+    #[error("No source vote found for cert {0:?} return_primary {1}")]
+    NoSourceVoteFound(CertificateType, bool),
+}
+
+pub struct BLSSigVerifier {
+    verified_votes_sender: VerifiedVoterSlotsSender,
+    message_sender: Sender<ConsensusMessage>,
+    sharable_banks: SharableBanks,
+    stats: BLSSigVerifierStats,
+    verified_certs: RwLock<HashSet<CertificateType>>,
+    vote_payload_cache: RwLock<HashMap<Vote, Arc<Vec<u8>>>>,
+    consensus_metrics_sender: ConsensusMetricsEventSender,
+    last_checked_root_slot: Slot,
+}
+
+impl BLSSigVerifier {
+    pub fn new(
+        sharable_banks: SharableBanks,
+        verified_votes_sender: VerifiedVoterSlotsSender,
+        message_sender: Sender<ConsensusMessage>,
+        consensus_metrics_sender: ConsensusMetricsEventSender,
+    ) -> Self {
+        Self {
+            sharable_banks,
+            verified_votes_sender,
+            message_sender,
+            stats: BLSSigVerifierStats::new(),
+            verified_certs: RwLock::new(HashSet::new()),
+            vote_payload_cache: RwLock::new(HashMap::new()),
+            consensus_metrics_sender,
+            last_checked_root_slot: 0,
+        }
+    }
+
+    // We use the custom pool in tests, global pool in production.
+    fn run_batch_verify_in_pool<F, R>(pool: Option<&ThreadPool>, f: F) -> R
+    where
+        F: FnOnce() -> R + Send,
+        R: Send,
+    {
+        match pool {
+            Some(pool) => pool.install(f),
+            None => f(),
+        }
+    }
+
+    fn purge_old_votes_and_certs(&mut self, root_slot: Slot) {
+        if self.last_checked_root_slot < root_slot {
+            self.last_checked_root_slot = root_slot;
+            self.verified_certs
+                .write()
+                .unwrap()
+                .retain(|cert| cert.slot() > root_slot);
+            self.vote_payload_cache
+                .write()
+                .unwrap()
+                .retain(|vote, _| vote.slot() > root_slot);
+        }
+    }
+
+    fn preprocess_vote(
```

Comment:
function header would be good here to indicate what kind of processing is happening